test/unit/atomic_iteration_test.lua

```diff
@@ -0,0 +1,165 @@
+        helpers.retrying({}, function()
+            if space.index[0].type == "HASH" then
+                t.assert_equals(helpers.iteration_result, {{3, "1"}, {2, "2"}, {1, "3"}})
+            else
+                t.assert_equals(helpers.iteration_result, {{1, "3"}, {2, "2"}, {3, "1"}})
+            end
+        end)
+        task:kill()
+        helpers.iteration_result = {}
+
+        -- check out three row transaction
+        if space.index[0].type == "HASH" then
+            t.assert_equals(transactions, {
+                { {3, "1"}, {2, "2"}, {1, "3"} }
+            })
+        else
+            t.assert_equals(transactions, {
+                { {1, "3"}, {2, "2"}, {3, "1"} }
+            })
+        end
+        transactions = {}
+
+        -- tuples expired in two atomic_iteration
+        space:insert({1, "3"})
+        space:insert({2, "2"})
+        space:insert({3, "1"})
+
+        task = expirationd.start("clean_all", space.id, helpers.is_expired_debug,
+                {atomic_iteration = true, tuples_per_iteration = 2})
+        -- wait for tuples expired
+        -- 2 seconds because suspend will be yield in task
+        helpers.retrying({}, function()
+            if space.index[0].type == "HASH" then
+                t.assert_equals(helpers.iteration_result, {{3, "1"}, {2, "2"}, {1, "3"}})
+            else
+                t.assert_equals(helpers.iteration_result, {{1, "3"}, {2, "2"}, {3, "1"}})
+            end
+        end)
+        task:kill()
+        helpers.iteration_result = {}
+
+        if space.index[0].type == "HASH" then
+            t.assert_equals(transactions, {
+                { {3, "1"}, {2, "2"} }, -- check two row transaction
+                { {1, "3"} }            -- check single row transactions
+            })
+        else
+            t.assert_equals(transactions, {
+                { {1, "3"}, {2, "2"} }, -- check two row transaction
+                { {3, "1"} }            -- check single row transactions
+            })
+        end
+
+        transactions = {}
+    end
+
+    -- unmock
+    box.begin = true_box_begin
+end
+
+function g.test_mvcc_vinyl_tx_conflict()
+    for i = 1,10 do
+        g.vinyl:insert({i, tostring(i), nil, nil, 0})
+    end
+
+    local updaters = {}
+    for i = 1,10 do
+        local updater = fiber.create(function()
+            fiber.name(string.format("updater of %d", i), { truncate = true })
+            while true do
+                g.vinyl:update({i}, { {"+", 5, 1} })
+            end
+        end)
+        table.insert(updaters, updater)
+    end
+
+    local task = expirationd.start("clean_all", g.vinyl.id, helpers.is_expired_debug,
+            {atomic_iteration = true})
+
+    -- wait for tuples expired
+    fiber.sleep(3)
+
+    task:kill()
+    for i = 1,10 do
+        fiber.kill(updaters[i]:id())
+    end
+
+    t.assert(box.stat.vinyl().tx.conflict > 0)
+    t.assert_equals(box.stat.vinyl().tx.conflict, box.stat.vinyl().tx.rollback)
```

Comment:
And seems we also should check that task still works and after there is no load anymore tuples are expired.

Reply:
And probably let's check `box.stat.vinyl().tx.transactions == 0` - there are no opened transactions.

Reply:
Fixed